django/db/models/query.py

```diff
@@ -1159,6 +1159,9 @@ def update(self, **kwargs):
         self._for_write = True
         query = self.query.chain(sql.UpdateQuery)
         query.add_update_values(kwargs)
+        query.order_by = tuple(
+            col for col in query.order_by if col not in query.annotations
+        )
```

Comment:
One thing that makes me uneasy here is that this code would make the following preserve ordering?

```
Bar.objects.order_by(Count("m2m_foo"))
```

Should we inline ordering references in ordering instead of silently eliminating them? It feels like we should as `annotate(foo=bar).order_by('foo')` should be equivalent to `order_by(bar)`?

Reply:
Alternatively, would it make sense/be viable to entirely drop the `order_by` as suggested Mariusz in https://code.djangoproject.com/ticket/28897#comment:6 with `query.clear_ordering()`?
The behaviour would at least be deterministic, without corner cases (and it would require some additional documentation maybe).

However, I'm not sure if we need to keep the order by for an `update` call, or if we can always prune it 🤔

Reply:
Hmm, actually `Bar.objects.order_by(Count("m2m_foo"))` is already raising an exception:
```
ERROR: test_update_ordered_by_m2m_annotation (update.tests.AdvancedTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/david/django/tests/update/tests.py", line 237, in test_update_ordered_by_m2m_annotation
    Bar.objects.order_by(Count("m2m_foo")).update(x=3)
  File "/home/david/django/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/david/django/django/db/models/query.py", line 1623, in order_by
    obj.query.add_ordering(*field_names)
  File "/home/david/django/django/db/models/sql/query.py", line 2206, in add_ordering
    raise FieldError(
django.core.exceptions.FieldError: Using an aggregate in order_by() without also including it in annotate() is not allowed: Count(F(m2m_foo))
```
However, that would only be true for aggregations.

You are right for
```
Bar.objects.order_by(Abs("m2m_foo"))
```
and
```
Bar.objects.annotate(abs_id=Abs("m2m_foo")).order_by("abs_id")
```
which should have the same behaviour but have not with the current code 🤔 

-------------------

Raising an error in the appropriate case would probably make sense. Here the different cases (when we call `.update` on them):
- inline aggregation `Bar.objects.order_by(Count("m2m_foo"))` => existing raise _Using an aggregate in order_by() without also including it in annotate() is not allowed_
- annotate aggregation `Bar.objects.annotate(m2m_count=Count("m2m_foo")).order_by("m2m_count")` => **new error**, something like _Cannot update when ordering on an annotation. Try using inline order_by(Count(F(m2m_foo)))_
- inline function `Bar.objects.order_by(Abs("m2m_foo"))` => should work
- annotate function `Bar.objects.annotate(abs_id=Abs("m2m_foo")).order_by("abs_id")` => **new error**, something like _Cannot update when ordering on an annotation. Try using inline order_by(Abs(F(m2m_foo)))_

However, the user experience is probably not great for two reasons:
1. when using an annotate aggregation, the message suggests that it might work when changing the caller - which is not true (and we know it)
2. when using an annotate function, we actually could automatically translate it.

I'll try to push changes for the "more advanced" version